src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs

```diff
@@ -224,36 +225,48 @@ public async Task<Diagnostic> ToDiagnosticAsync(Project project, CancellationTok
                 location, additionalLocations, customTags: CustomTags, properties: Properties);
         }
 
-        public static TextSpan GetTextSpan(DiagnosticDataLocation? dataLocation, SourceText text)
+        public static (LinePosition startLine, LinePosition endLine) GetLinePositions(DiagnosticDataLocation? dataLocation, SourceText text, bool useMapped)
         {
             var lines = text.Lines;
             if (lines.Count == 0)
             {
                 return default;
             }
 
-            var originalStartLine = dataLocation?.OriginalStartLine ?? 0;
-            if (originalStartLine >= lines.Count)
+            var dataLocationStartLine = (useMapped ? dataLocation?.MappedStartLine : dataLocation?.OriginalStartLine) ?? 0;
+            var dataLocationStartColumn = (useMapped ? dataLocation?.MappedStartColumn : dataLocation?.OriginalStartColumn) ?? 0;
+            var dataLocationEndLine = (useMapped ? dataLocation?.MappedEndLine : dataLocation?.OriginalEndLine) ?? 0;
+            var dataLocationEndColumn = (useMapped ? dataLocation?.MappedEndColumn : dataLocation?.OriginalEndColumn) ?? 0;
+
+            if (dataLocationStartLine >= lines.Count)
```

Comment:
how could this possibly happen?

Reply:
Not sure - just moved the logic.  theoretically, if the diagnostic was published for a different version of the text, it could be outside the bounds?

Reply:
>  theoretically, if the diagnostic was published for a different version of the text,

When does that happen?  In roslyn, everything should be snapshotted.  So it shouldn't be possible to have diagnostics published for different versions right?  Does LSP change that?

Reply:
LSP doesn't change that, just going off of this comment - http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs,202

Reply:
:( :( :( :( :( :( :( :( :( :( 

Ok... :(

I guess we have to be safe from this case then.

Reply:
Do we have a tracking bug for that cleanup though? The internal bug in that code regarding the NFW is two and a half years old...

Reply:
I think the internal bug is referring to something else.  The link I posted was to the comment that was expecting the diagnostic to be published for a different document version, and the corresponding mapping code to ensure it's in bounds for the current version.

Reply:
> The link I posted was to the comment that was expecting the diagnostic to be published for a different document version, and the corresponding mapping code to ensure it's in bounds for the current version.

It's super weird though... so we have a diagnostic for a span that seems to no longer exist... but we're trying to map that span to our current doc.  

My pref would be to just drop the diagnostic on the floor then.  

--

However, if htis is what we already do, i guess i'm ok with it.  I just don't like that the actual design here is just "we have snapshots.  we don't use them... but we have them.  and here's some data about some random version that you have to reconcile with whatever actual version you have".

It's basically broken by-design and doesn't follow *any* of our other snapshot-oriented systems we have in roslyn.